vertical-pod-autoscaler/updater/eviction/pods_eviction_restriction.go

```diff
@@ -0,0 +1,236 @@
+	}
+	if e.evictionBudget[cr] < 1 {
+		return fmt.Errorf("cannot evict pod %v : eviction budget exceeded", podToEvict.Name)
+	}
+
+	eviction := &policyv1.Eviction{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: podToEvict.Namespace,
+			Name:      podToEvict.Name,
+		},
+	}
+	err := e.client.CoreV1().Pods(podToEvict.Namespace).Evict(eviction)
+	if err != nil {
+		glog.Errorf("failed to evict pod %s, error: %v", podToEvict.Name, err)
+		return err
+	}
+	e.evictionBudget[cr] = e.evictionBudget[cr] - 1
+	return nil
+}
+
+// NewPodsEvictionRestrictionFactory creates PodsEvictionRestrictionFactory
+func NewPodsEvictionRestrictionFactory(client kube_client.Interface, minReplicas int, evictionToleranceFraction float64) PodsEvictionRestrictionFactory {
+	return &podsEvictionRestrictionFactoryImpl{client: client, minReplicas: minReplicas, evictionToleranceFraction: evictionToleranceFraction}
+}
+
+// NewPodsEvictionRestriction creates PodsEvictionRestriction for given set of pods.
+func (f *podsEvictionRestrictionFactoryImpl) NewPodsEvictionRestriction(pods []*apiv1.Pod) PodsEvictionRestriction {
+	// We can evict pod only if it is a part of replica set
+	// For each replica set we can evict only a fraction of pods.
+	// Evictions may be later limited by pod disruption budget if configured.
+
+	livePods := make(map[podReplicaCreator][]*apiv1.Pod)
+
+	for _, pod := range pods {
+		creator, err := getPodReplicaCreator(pod)
+		if err != nil {
+			glog.Errorf("failed to obtain replication info for pod: %v", pod.Name, err)
+			continue
+		}
+		if creator == nil {
+			glog.V(2).Infof("pod not replicated %v", pod.Name)
+			continue
+		}
+		livePods[*creator] = append(livePods[*creator], pod)
+	}
+
+	podsCreators := make(map[string]podReplicaCreator)
+	creatorsEvictionBudget := make(map[podReplicaCreator]int)
+	for creator, replicas := range livePods {
+		actual := len(replicas)
+		if actual < f.minReplicas {
+			glog.V(2).Infof("too few replicas for %v %v/%v. Found %v live pods",
+				creator.Kind, creator.Namespace, creator.Name, actual)
+			continue
+		}
+		var configured int
+		if creator.Kind == "Job" {
+			// Job has no replicas configuration, so we will use actual number of live pods as replicas count.
+			configured = actual
+		} else {
+			var err error
+			configured, err = getReplicaCount(creator, f.client)
+			if err != nil {
+				glog.Errorf("failed to obtain replication info for %v %v/%v. %v",
+					creator.Kind, creator.Namespace, creator.Name, err)
+				continue
+			}
+		}
+
+		evictionTolerance := int(float64(configured) * f.evictionToleranceFraction)
+		currentlyEvicted := configured - actual
+		evictionBudget := evictionTolerance - currentlyEvicted
+
+		if evictionBudget > 0 {
+			creatorsEvictionBudget[creator] = evictionBudget
+		} else {
+			creatorsEvictionBudget[creator] = 1
+			glog.V(2).Infof("eviction tolerance too low for pods from %v %v/%v. Eviction budget limit : found %v live pods, configured replicas %v",
```

Comment:
This debug is wrong. It doesn't convey the fact that we increased the allowed evictions to 1.